lib/cdo/poste.rb

```diff
@@ -105,6 +107,242 @@ def self.unsubscribe(email, hashed_email, params={})
       )
     end
   end
+
+  class Template
+    def initialize(body, engine=TextRender::MarkdownEngine)
+      if match = body.match(/^---\s*\n(?<header>.*?\n?)^(---\s*$\n?)(?<html>\s*\n.*?\n?)^(---\s*$\n?)(?<text>\s*\n.*?\n?\z)/m)
+        @header = TextRender::YamlEngine.new(match[:header].strip)
+        @html = engine.new(match[:html].strip)
+        @text = TextRender::ErbEngine.new(match[:text].strip)
+      elsif match = body.match(/^---\s*\n(?<header>.*?\n?)^(---\s*$\n?)(?<html>\s*\n.*?\n?\z)/m)
+        @header = TextRender::YamlEngine.new(match[:header].strip)
+        @html = engine.new(match[:html].strip)
+      else
+        @html = engine.new(body.strip)
+      end
+    end
+
+    def render(params={})
+      if params.key?('form_id')
+        form = Form2.from_row(DB[:forms].where(id: params['form_id']).first)
+        params.merge! form.data
+        params.merge! form.processed_data
+        params['form'] = form
+      end
+      locals = OpenStruct.new(params).instance_eval {binding}
+
+      header = @header.result(locals) unless @header.nil?
+      # TODO(andrew): Fix this so that we get a signal as to how often this is happening.
+      # For more information, see https://www.pivotaltracker.com/story/show/104750788.
+      tracking_id = header['litmus_tracking_id'] unless header.nil?
+
+      html = @html.result(locals) if @html
+      # Parse the html into a DOM and then re-serialize back to html text in case we were depending on that
+      # logic in the click tracking method to clean up or canonicalize the HTML.
+      html = Nokogiri::HTML(html).to_html if html
+      html = inject_litmus_tracking html, tracking_id, params[:encrypted_id] if html
+      text = @text.result(locals) unless @text.nil?
+
+      [header, html, text]
+    end
+
+    def inject_litmus_tracking(html, tracking_id, unique_id)
+      return html unless tracking_id && unique_id
+      litmus_blob = <<-eos
+<style>@media print{ #_t { background-image: url('https://#{tracking_id}.emltrk.com/#{tracking_id}?p&d=#{unique_id}');}} div.OutlookMessageHeader {background-image:url('https://#{tracking_id}.emltrk.com/#{tracking_id}?f&d=#{unique_id}')} table.moz-email-headers-table {background-image:url('https://#{tracking_id}.emltrk.com/#{tracking_id}?f&d=#{unique_id}')} blockquote #_t {background-image:url('https://#{tracking_id}.emltrk.com/#{tracking_id}?f&d=#{unique_id}')} #MailContainerBody #_t {background-image:url('https://#{tracking_id}.emltrk.com/#{tracking_id}?f&d=#{unique_id}')}</style><div id="_t"></div>
+<img src="https://#{tracking_id}.emltrk.com/#{tracking_id}?d=#{unique_id}" width="1" height="1" border="0" />
+eos
+      html.gsub("</body>", litmus_blob + "\n</body>")
+    end
+  end
+end
+
+class Deliverer
```

Comment:
Is there a reason not to give each class its own file?

Reply:
mostly a desire not to clutter up the already-large `lib/cdo` directory